documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,824 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
 
-The API surface for this is defined later in the Proposed API section.
+The `CreateInput` methods simply return an `InputContext` preconfigured with the backend created by `CreateInputBackend` for ease of use.
 
-One instance of an `IInputBackend` can only belong to one `InputContext` throughout the entirety of the `IInputBackend`'s lifetime. Once a `IInputBackend` is `Add`ed to an `InputContext`, the `InputContext` takes full ownership of the `IInputBackend` and, as a result, will `Dispose` the `IInputBackend` when it is `Remove`d or the context itself is `Dispose`d. If multiple `InputContext`s are in use, multiple `IInputBackend` instances must also be used. `IInputBackend` will throw an exception if its APIs are used after the backend has been disposed.
+Please see the Windowing 3.0 proposal for `IWindowHandlesSource` and `WindowHandles`.
 
-**KEY POINT FOR WORKING GROUP**: The Windowing-Input integration mandates the use of source generators. Is this ok?
+# Devices
 
-# Proposed API
+Input devices all inherit from a root interface.
 
-```diff
-namespace Silk.NET.Input
+```cs
+public interface IInputDevice
 {
-    public interface IInputDevice
-    {
--       int Index { get; }
-+       /// <summary>
-+       /// The backend-specific, device-type-specific identifier for this device.
-+       /// </summary>
-+       /// <remarks>
-+       /// For example, no gamepad may share a DeviceId with another gamepad (same applies for all other device types) on that individual backend.
-+       /// This property should not be used as a globally unique identifier.
-+       /// </remarks>
-+       int DeviceId { get; }
-     }
+    nint Id { get; }
+    string Name { get; }
 }
 ```
 
-Index has been removed in favour of DeviceId, as Index implies its globally unique and also tied to the index of the device in the list in which it's contained. The reason why this is not globally unique is primarily to allow users to do their own interop with the native backend using our high-level representations of the devices (i.e. DeviceId will be the GLFW joystick ID)
+`Id` is an globally-unique integral identifier for this device.
+
+`Name` is a rough description of the input device. Its value is not intrinsically meaningful.
+
+All devices originate from a backend.
+
+# Backends
 
-```diff
-namespace Silk.NET.Input
+```cs
+public interface IInputBackend
 {
--   public interface IInputPlatform
--   {
--       bool IsApplicable(IView view);
--       IInputContext CreateInput(IView view);
--   }
+    string Name { get; }
+    nint Id { get; }
+    IReadOnlyList<IInputDevice> Devices { get; }
+    void Update(IInputHandler? actor = null);
```

Comment:
Could the `Update(handler)` logic bring performance disadvantages? It has to check whether the passed IInputHandler also implements the other interfaces. What if instead we pass this object once in something like `IInputBacked.SetInputHandler()` so these checks don't have to happen all the time?

Reply:
If they do, it'd only be until PGO kicks in and in any case I'm confident that the performance impact will be negligible: a few type checks won't touch a ~360FPS game loop (which I'm pretty sure 360Hz is the highest refresh rate monitor around today, so I'm using that as the maximum meaningful FPS)

We can do testing of course, and could possibly even reserve the right to change this method if you'd like according to testing, but I'm not convinced that using the type system will cause significant performance impact for the amount we're calling it.

Reply:
For reference disassembly:
```cs

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static IInputBackend GetBackend() => new Backend();
    public static void M() {
        var myThing = GetBackend();
        var myOtherThing = new Nothing();
        myThing.Update((IInputHandler)myOtherThing);
    }
```
```asm

    L0000: push rsi
    L0001: sub rsp, 0x20
    L0005: mov rcx, 0x7ffb2751db20
    L000f: call 0x00007ffb7b5baf20
    L0014: mov rsi, rax
    L0017: call 0x00007ffb275104f0
    L001c: mov rcx, rax
    L001f: mov rdx, rsi
    L0022: mov r11, 0x7ffb27517000
    L002c: mov rax, [0x7ffb27517000]
    L0033: add rsp, 0x20
    L0037: pop rsi
    L0038: jmp rax
```

Not clean, but not the worst thing in the world for something that is called once per update in the typical case. And CPU caching might help us out further with those calls. I wouldn't be too worried.